packages/dds/tree/src/feature-libraries/indexing/anchorTreeIndex.ts

```diff
@@ -0,0 +1,417 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { assert } from "@fluidframework/core-utils/internal";
+import { disposeSymbol, fail, getOrCreate } from "../../util/index.js";
+import {
+	type Anchor,
+	type AnchorNode,
+	type FieldKey,
+	type TreeNodeSchemaIdentifier,
+	forEachField,
+	forEachNode,
+	type ITreeSubscriptionCursor,
+	createAnnouncedVisitor,
+	type IForestSubscription,
+	type AnnouncedVisitor,
+	CursorLocationType,
+	rootField,
+	type UpPath,
+	type ProtoNodes,
+	keyAsDetachedField,
+} from "../../core/index.js";
+import type { TreeIndex, TreeIndexKey, TreeIndexNodes } from "./types.js";
+import { TreeStatus } from "../flex-tree/index.js";
+
+/**
+ * A function that gets the value to index a node on.
+ * The given cursor should point to the node that will be indexed.
+ *
+ * @returns a value on the node that the index will use as
+ *
+ * @remarks
+ * This function does not own the cursor in any way, it walks the cursor to find the key the node is indexed on
+ * but returns the cursor to the state it was in before being passed to the function. It should also not be disposed by this function
+ * and must be disposed elsewhere.
+ */
+export type KeyFinder<TKey extends TreeIndexKey> = (tree: ITreeSubscriptionCursor) => TKey;
+
+/**
+ * An index from some arbitrary keys to anchor nodes. Keys can be anything that is a {@link TreeValue}.
+ * A key can map to multiple nodes but each collection of nodes only results in a single value.
+ *
+ * @remarks
+ * Detached nodes are stored in the index but filtered out when any public facing apis are called. This means that
+ * calling {@link keys} will not include any keys that are stored in the index but only map to detached nodes.
+ */
+export class AnchorTreeIndex<TKey extends TreeIndexKey, TValue>
+	implements TreeIndex<TKey, TValue>
+{
+	/**
+	 * Caches {@link KeyFinder}s for each schema definition. If a schema maps to null, it does not
+	 * need to be considered at all for this index. This allows us to skip subtrees that aren't relevant
+	 * as a performance optimization.
+	 */
+	private readonly keyFinders = new Map<TreeNodeSchemaIdentifier, KeyFinder<TKey> | null>();
+	/**
+	 * The actual index from keys to anchor nodes.
+	 */
+	private readonly nodes = new Map<TKey, AnchorNode[]>();
+	/**
+	 * Maintained for efficient removal of anchor nodes from the index when updating their keys
+	 */
+	private readonly anchorKeys = new Map<AnchorNode, TKey>();
+	/**
+	 * Keeps track of anchors for disposal.
+	 */
+	private readonly anchors = new Map<AnchorNode, Anchor[]>();
+
+	/**
+	 * @param forest - the forest that is being indexed
+	 * @param indexer - a function that retrieves the key finder based on a given schema or undefined if the schema does not have an associated key finder
+	 * @param getValue - a function that returns the value or undefined given at least one anchor node
```

Comment:
must also be functional and pure

Reply:
also, this doc doesn't really tell me what this is for. it's a bit of an odd abstraction (keyFinder is intuitive, but this getValue isn't IMO) so adding something would be helpful